components/utils.py

```diff
@@ -170,6 +174,93 @@ def remove(path, ignore_failure=False):
                         .format(path))
 
 
+def _generate_ssl_cert(cert_filename, key_filename, cn):
+
+    alt_name = 'IP:{0}'.format(cn)
+
+    fd, conf_file = tempfile.mkstemp()
+    os.close(fd)
+    with open(conf_file, 'w') as f:
+        f.write('subjectAltName={0}\n'
+                'distinguished_name = req_distinguished_name\n'
+                '[ req_distinguished_name ]'
+                'commonName={1}'.format(alt_name, cn))
+
+    # building openssl command
+    command_arr = shlex.split('openssl req -x509 -nodes -newkey rsa:2048 '
+                              '-out {0} -keyout {1} -days 3650 -batch '
+                              '-subj \'/CN={2}\' -config {3}'.
+                              format(cert_filename, key_filename, cn,
+                                     conf_file))
+    sudo(command_arr)
+    os.remove(conf_file)
+
+
+def deploy_ssl_cert_and_key(cert_filename, key_filename, cn):
+    """
+    SSL certificate and keys can be supplied by the user in the
+    manager-blueprint 'resources/ssl' directory.
+    If SSL certs were supplied - Cloudify will use them,
+    otherwise - they will be generated
+    :param cert_filename: the name of the certificate file (e.g. rest.crt)
+    :param key_filename: the name of the key file (e.g. rest.key)
+    :param cn: the common name to use for certificates creation (e.g.
+        myserver.com)
+    """
+    mkdir(SSL_CERTS_TARGET_DIR)
+    user_supplied_cert_path = \
+        os.path.join(SSL_CERTS_SOURCE_DIR, cert_filename)
+    user_supplied_key_path = \
+        os.path.join(SSL_CERTS_SOURCE_DIR, key_filename)
+    cert_target_path = os.path.join(SSL_CERTS_TARGET_DIR, cert_filename)
+    key_target_path = os.path.join(SSL_CERTS_TARGET_DIR, key_filename)
+    try:
+        # trying to deploy pre-existing certificates
+        ctx.logger.info('Deploying SSL certificate "{0}" and SSL private '
+                        'key "{1}"...'.format(cert_filename, key_filename))
+        deploy_blueprint_resource(user_supplied_cert_path,
+                                  cert_target_path,
+                                  NGINX_SERVICE_NAME,
+                                  user_resource=True,
+                                  load_ctx=False)
+        deploy_blueprint_resource(user_supplied_key_path,
+                                  key_target_path,
+                                  NGINX_SERVICE_NAME,
+                                  user_resource=True,
+                                  load_ctx=False)
+    except subprocess.CalledProcessError as e:
+        if "No such file or directory" in e.stderr:
```

Comment:
Actually instead of examining the exception message, we can check if the certificate now exists on the manager (in the `except` block or even always)